packages/rum/src/boot/recorderApi.ts

```diff
@@ -186,6 +166,38 @@ export function makeRecorderApi(
         }
       }
 
+      const doStart = async () => {
+        const [startRecordingImpl] = await Promise.all([
+          loadRecorder(),
+          asyncRunOnReadyState(configuration, 'interactive'),
+        ])
+
+        if (state.status !== RecorderStatus.Starting) {
+          return
+        }
+
+        const deflateEncoder = getOrCreateDeflateEncoder()
+        if (!deflateEncoder) {
+          state = {
+            status: RecorderStatus.Stopped,
+          }
+          return
+        }
+
+        const { stop: stopRecording } = startRecordingImpl(
+          lifeCycle,
+          configuration,
+          sessionManager,
+          viewHistory,
+          deflateEncoder
+        )
+
+        state = {
+          status: RecorderStatus.Started,
+          stopRecording,
+        }
+      }
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏ As we're using promise, I'll avoid having state all over the place. maybe something like this would be a bit more clean:


```ts
const doStart = async () => {
  const [startRecordingImpl] = await Promise.all([
    loadRecorder(),
    asyncRunOnReadyState(configuration, 'interactive'),
  ])

  if (state.status !== RecorderStatus.Starting) {
    return
  }

  const deflateEncoder = getOrCreateDeflateEncoder()
  if (!deflateEncoder) {
    throw new Error('deflate encoder not initialized!');
  }

  const { stop: stopRecording } = startRecordingImpl(
    lifeCycle,
    configuration,
    sessionManager,
    viewHistory,
    deflateEncoder
  )
  
  return stopRecording
}
```

And manage the state when you use it:
```ts
 doStart()
  .then(stop => {
    state = {
      status: RecorderStatus.Started,
      stopRecording: stop
    }
  })
  .catch(() => {
    state = {
      status: RecorderStatus.Stopped,
    }
  })
```

Reply:
Interesting! I'll look at it

Reply:
I'm not a big fan of throwing errors for control flow. Typescript does not have a syntax to tell whether a function can throw.

Having a `Result` type could be nicer ([ex](https://imhoff.blog/posts/using-results-in-typescript)), as it can be typechecked properly.

Or, you can make `doStart` work as a state machine and return the new state:

```ts
     const doStart = async () => {
        const [startRecordingImpl] = await Promise.all([
          loadRecorder(),
          asyncRunOnReadyState(configuration, 'interactive'),
        ])

        if (state.status !== RecorderStatus.Starting) {
          return state
        }

        const deflateEncoder = getOrCreateDeflateEncoder()
        if (!deflateEncoder) {
          return {
            status: RecorderStatus.Stopped,
          }
        }

        const { stop: stopRecording } = startRecordingImpl(
          lifeCycle,
          configuration,
          sessionManager,
          viewHistory,
          deflateEncoder
        )

        return {
          status: RecorderStatus.Started,
          stopRecording,
        }
      }
```

Not sure if it simplifies things though.

Reply:
I've worked on a refacto for the recorderApi to mimic what we have in RumPublicAPI:
Having preStart and postStart strategies. I'll create a dedicated PR for it.

Reply:
https://github.com/DataDog/browser-sdk/pull/3181

Reply:
We can use the promise chain as mentioned as it would make two async tasks. One starting the recorder and another changing the status.

Reply:
With this refacto PR, things should be clearer: https://github.com/DataDog/browser-sdk/pull/3181